omnilib/control.lua

```diff
@@ -1,408 +1,343 @@
+					if updated_ingredients ~= 0 then
+						log("\t\tMigrated " .. updated_ingredients .. " ingredients")
 					end
 				end
 			end
 		end
 	end
-	log("omnidate completed")
-	initilizing=false
+	--log("Building update complete")
+end
+
+local initializing = false
+function omni_update(game, silent)
+	initializing = true
+	--log("Beginning omnidate")
+	local profiler = game.create_profiler()
+	for _, force in pairs(game.forces) do
+		update_force(force)
+	end
+	update_building_recipes()
+	if not silent then
+		game.print({
+			"",
+			"Omnidate completed. ",
+			profiler
+		})
+	else
+		log({
+			"",
+			"Omnidate completed. ",
+			profiler
+		})
+	end
+	initializing = false
 end
 
 script.on_event(defines.events.on_console_chat, function(event)
-	if event.player_index then
-		local player = game.players[event.player_index]
+	if event.player_index and game.players[event.player_index] then
 		if event.message=="omnidate" then
-		omni_update(game)
+			omni_update(game)
 		end
 	end
 end)
 
-local omni_cat={}
-
-function acquireData(game)
-	local perm = {}
-	for _, rec in pairs(game.recipe_prototypes) do
-		if not omni_cat[rec.category] then
-			omni_cat[rec.category]={name=rec.category,buildings={}}
-		end
-		if string.find(rec.name,"omniperm") then
-			local s,e = string.find(rec.name,"omniperm")
-			local v = split(string.sub(rec.name,e+2,string.len(rec.name)),"-")
-			local name = string.sub(rec.name,1,s-2)
-			if not perm[name] then perm[name]={ingredient=1,result=1} end
-			if tonumber(v[1])>perm[name].ingredient then perm[name].ingredient = tonumber(v[1]) end
-			if tonumber(v[2])>perm[name].result then perm[name].result = tonumber(v[2]) end
-		end
-	end
-	for name,entity in pairs(game.entity_prototypes) do
-		if entity.crafting_categories and not string.find(name,"character") then
-			for i, c in pairs(entity.crafting_categories) do
-				if omni_cat[i] then
-					table.insert(omni_cat[i].buildings,entity.name)
-				end
+local patterns = {
+    "%-()%a$",
+    "%-()%a%-omniperm",
+    "%-()%a%-compression"
+}
+function acquire_data(game)
+	-- Recipe tier map
+	local recipes = {}
+	local recipe_map = {}
+	for name, recipe in pairs(game.recipe_prototypes) do
+		for index, pattern in pairs(patterns) do
+			local match = select(3,name:find(pattern))-- We only care about match position, denoted by ()
+			if match and recipe.category and recipe.category:find("omni") then-- KR2 recipe names will screw us otherwise
+				local base = name:sub(1, match-1)
+				local tier = name:sub(match, match):byte()
+				local variant = name:sub(match+1)
+				recipes[name] = {
+					base = base, -- The entry name in the common table
+					tier = tier,
+					variant = variant,
+					compressed = (index == 3)
+				}
+				recipe_map[base] = recipe_map[base] or {}
+				recipe_map[base].active_tier = math.min(97, recipe_map[base].active_tier or tier)
+				recipe_map[base][tier] = recipe_map[base][tier] or {}				
+				recipe_map[base][tier][variant] = name
+				break
```

Comment:
A lot is happening here, so I'll explain:
We check every recipe against our recipe-name-tierletter-modifier pattern. The first pattern indicates a recipe has no modifiers, AKA it is not compressed or permuted. The second is for permutations, and the third will indicate if the recipe is compressed.

From here, our data scheme has two parts. We have recipe_map, which keeps our current state, and recipes which keeps the keys and metadata we use to lookup items in recipe_map.

Recipes are indexed by prefix, then by tier as the letter in bytes, then by suffix. An example dump from just omnimatter/omnipermute is below.

https://gist.github.com/oorzkws/67f0599a7968c3b3ced91e689d8b37ff